File: datasets/Python/Gemini/Algorithm/hard5.py
import sys

def solve():
    """
    Calculates the number of winning Nim games (ordered tuples) with n distinct piles,
    each with < 2^m stones.
    The number of winning games equals the number of ways to choose n linearly 
    independent vectors from GF(2)^m.
    """
    try:
        n_str, m_str = sys.stdin.readline().split()
        n = int(n_str)
        m = int(m_str)
    except (IOError, ValueError):
        return

    MOD = 10**9 + 7

    def power(base, exp):
        res = 1
        base %= MOD
        while exp > 0:
            if exp % 2 == 1:
                res = (res * base) % MOD
            base = (base * base) % MOD
            exp //= 2
        return res

    # If n > 2^m, it's impossible to choose n distinct vectors,
    # but the loop condition handles this naturally as (2^m - 2^i) will become 0.
    
    p2m = power(2, m)
    winning_games = 1
    p2i = 1  # This will hold 2^i

    for i in range(n):
        term = (p2m - p2i + MOD) % MOD
        winning_games = (winning_games * term) % MOD
        p2i = (p2i * 2) % MOD
        
    print(winning_games)

solve()

File: datasets/C++/Gemini/Algorithm/hard5.cpp
#include <iostream>
#include <vector>

using namespace std;

long long power(long long base, long long exp) {
    long long res = 1;
    long long MOD = 1000000007;
    base %= MOD;
    while (exp > 0) {
        if (exp % 2 == 1) res = (res * base) % MOD;
        base = (base * base) % MOD;
        exp /= 2;
    }
    return res;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    long long m;
    cin >> n >> m;

    long long MOD = 1000000007;

    long long p2m = power(2, m);
    long long winning_games = 1;
    long long p2i = 1; // This will hold 2^i

    for (int i = 0; i < n; ++i) {
        long long term = (p2m - p2i + MOD) % MOD;
        winning_games = (winning_games * term) % MOD;
        p2i = (p2i * 2) % MOD;
    }

    cout << winning_games << endl;

    return 0;
}